### PR TITLE
[Feature] Support RTMDet and RTMPose ncnn deployment

### PR DESCRIPTION
**Pull request by hanrui1sensetime**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support RTMDet and RTMPose ncnn deployment

## Modification

- [x] Add rewriter for rtmpose_block.
- [x] Add configs.

## BC-breaking (Optional)

None.

## Use cases (Optional)

None.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdeploy/codebase/mmpose/models/utils/rtmpose_block.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import torch
+import torch.nn.functional as F
+from mmpose.models.utils import rope
+
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmpose.models.utils.rtmpose_block.ScaleNorm.forward', backend='ncnn')
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/utils/rtmpose_block.py:11]: Which mmpose branch are you using?

Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmpose/models/utils/rtmpose_block.py:11]: > Which mmpose branch are you using?

Now is mmpose dev-1.x branch.

File: mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmpose.models.utils.rtmcc_block.ScaleNorm.forward', backend='ncnn')
+def scalenorm__forward__ncnn(self, x):
+    """Rewrite `scalenorm` for ncnn backend.
+
+    ncnn does not support negative dimension for torch.chunk and torch.cat ncnn
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py:15]: Where is chunk and cat in this function?

File: mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    pad shape does not support float input
+    """
+    # The one-dim of Fubinious norm is equal to L2Norm.
+    # Set p=2 explicitly to map torch.norm to ReduceL2 onnx op,
+    # which will avoid FP16 exceed.
+    norm = torch.norm(x, p=2, dim=2, keepdim=True)
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py:21]: Why don't we rewrite torch.norm?

File: mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    norm = norm.clamp(min=self.eps)
+    return (x.unsqueeze(2) / norm.unsqueeze(2)).squeeze(2) * self.g
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmpose.models.utils.rtmcc_block.RTMCCBlock._forward', backend='ncnn')
+def rtmccblock___forward_ncnn(self, inputs):
+    """Rewrite `_forward` of RTMBlock for ncnn backend.
+
+    ncnn does not support negative dimension for Split op.
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py:33]: Why don't we rewite split and compute dim by the input shape?

File: mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        x = inputs
+    else:
+        x, k, v = inputs
+
+    x = self.ln(x)
+    uv = self.uv(x)
+    if self.attn_type == 'self-attn':
+        uv = self.act_fn(uv)
+        u = uv[..., :self.e]
+        v = uv[..., 512:1024]
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/utils/rtmcc_block.py:45]: magic number

File: mmdeploy/pytorch/functions/normalize.py [hunk 6]
```diff
@@ -39,3 +41,26 @@ def normalize__ncnn(input: torch.Tensor,
...
+    """Rewrite `torch.norm` for ncnn backend.
+
+    Rewrite torch.norm when p is Frobenius norm to avoid FP16 exceed in ncnn
+    Android platform.
+    """
+    ctx = FUNCTION_REWRITER.get_context()
+    origin_func = ctx.origin_func
+    if p == 'fro' and (isinstance(dim, int) or len(dim) == 1):
+        # Substitute Frobenius norm with L2 norm.
+        return torch.norm(
```
Comment (grimoire) [inline mmdeploy/pytorch/functions/normalize.py:62]: use origin_func here.

**Comment by grimoire** [comment]
visualize failed with 
```
2023-03-16:11:32:18 - root - ERROR - not enough values to unpack (expected 2, got 1)
Traceback (most recent call last):
  File "/home/user/space/repo2.0/mmdeploy/mmdeploy/utils/utils.py", line 41, in target_wrapper
    result = target(*args, **kwargs)
  File "/home/user/space/repo2.0/mmdeploy/mmdeploy/apis/visualize.py", line 72, in visualize_model
    result = model.test_step(model_inputs)[0]
  File "/home/user/space/repo2.0/mmengine/mmengine/model/base_model/base_model.py", line 145, in test_step
    return self._run_forward(data, mode='predict')  # type: ignore
  File "/home/user/space/repo2.0/mmengine/mmengine/model/base_model/base_model.py", line 326, in _run_forward
    results = self(**data, mode=mode)
  File "/home/user/anaconda3/envs/repo2.0/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user/space/repo2.0/mmdeploy/mmdeploy/codebase/mmpose/deploy/pose_detection_model.py", line 104, in forward
    batch_pred_x, batch_pred_y = batch_outputs
ValueError: not enough values to unpack (expected 2, got 1)

```

**Comment by hanrui1sensetime** [comment]
> visualize failed with
> 
> ```
> 2023-03-16:11:32:18 - root - ERROR - not enough values to unpack (expected 2, got 1)
> Traceback (most recent call last):
>   File "/home/user/space/repo2.0/mmdeploy/mmdeploy/utils/utils.py", line 41, in target_wrapper
>     result = target(*args, **kwargs)
>   File "/home/user/space/repo2.0/mmdeploy/mmdeploy/apis/visualize.py", line 72, in visualize_model
>     result = model.test_step(model_inputs)[0]
>   File "/home/user/space/repo2.0/mmengine/mmengine/model/base_model/base_model.py", line 145, in test_step
>     return self._run_forward(data, mode='predict')  # type: ignore
>   File "/home/user/space/repo2.0/mmengine/mmengine/model/base_model/base_model.py", line 326, in _run_forward
>     results = self(**data, mode=mode)
>   File "/home/user/anaconda3/envs/repo2.0/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/home/user/space/repo2.0/mmdeploy/mmdeploy/codebase/mmpose/deploy/pose_detection_model.py", line 104, in forward
>     batch_pred_x, batch_pred_y = batch_outputs
> ValueError: not enough values to unpack (expected 2, got 1)
> ```

May be I need your command to reproduce.

**Review by grimoire — APPROVED** [review]
LGTM

**Review by AllentDan — COMMENTED** [review]
The .cpp files need to be clang-formatted.

File: mmdeploy/codebase/mmdet/models/dense_heads/rtmdet_head.py [hunk 7]
```diff
@@ -105,3 +106,120 @@ def __mark_pred_maps(cls_scores, bbox_preds):
...
+                                              self.cls_out_channels)
+        for cls_score in cls_scores
+    ]
+    flatten_bbox_preds = [
+        bbox_pred.permute(0, 2, 3, 1).reshape(batch_size, -1, 4)
+        for bbox_pred in bbox_preds
+    ]
+
+    cls_scores = torch.cat(flatten_cls_scores, dim=1).sigmoid()
+    dummy_cls_scores = torch.zeros(
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/rtmdet_head.py:186]: Why do we need a dummy_cls_scores?

Comment (grimoire) [reply mmdeploy/codebase/mmdet/models/dense_heads/rtmdet_head.py:186]: Oh, I see, is it background class?

**Comment by hanrui1sensetime** [comment]
> The .cpp files need to be clang-formatted.

Will be formatted in another PR.

File: mmdeploy/codebase/mmdet/models/dense_heads/rtmdet_head.py (hunk 7 above)
Comment (hanrui1sensetime) [reply mmdeploy/codebase/mmdet/models/dense_heads/rtmdet_head.py:186]: > Oh, I see, is it background class?

Yes, it is for adapting NCNN DetectionOutput format.

**Review by AllentDan — APPROVED** [review]

**Pull request merged**